mmengine/utils/misc.py

```diff
@@ -217,6 +217,54 @@ def concat_list(in_list):
     return list(itertools.chain(*in_list))
 
 
+def apply_to(data: Any, expr: Callable, apply_func: Callable):
+    """Apply function to each element in dict, list or tuple that matches with
+    the expression.
+
+    For examples, if you want to convert each element in a list of dict from
+    `np.ndarray` to `Tensor`. You can use the following code:
+
+    Examples:
+
+    >>> from mmengine.utils import apply_to
+    >>> import numpy as np
+    >>> import torch
+    >>> data = dict(array=[np.array(1)]) # {'array': [array(1)]}
+    >>> result = apply_to(data, lambda x: isinstance(x, np.ndarray), lambda x: torch.from_numpy(x))
+    >>> print(result) # {'array': [tensor(1)]}
+
+    Args:
+        data (Any): Data to be applied.
+        expr (Callable): Expression to tell which data should be applied with
+            the function.
+        apply_func (Callable): Function applied to data.
+
+    Returns:
+        Any: The data after applying.
+    """  # noqa: E501
+    if any(map(expr, (dict(), '', b'', tuple(), list()))):
+        raise ValueError(
+            '``expr`` should not match with Mapping, str, bytes, tuple or '
+            'list. This kind of `expr` could lead to an unexpected results')
+
+    def _apply(data, expr, apply_func):
+        if isinstance(data, dict):
+            return {key: apply_to(data[key], expr, apply_func) for key in data}
```

Comment:
If calling `apply_to` rather than `_apply`, is it necessary to define the `_apple` inner function?

Reply:
`_apply` should call `_apply` rather than `apply_to`. I will fix it. Define `_apply` here is to avoid checking `expr` multiple times.

Reply:
Is it necessary to avoid checking expr? BTW, if decide to define `_apply` in the `apply_to`, there is no need to add `expr` and `apply_func` in its (_apply) signature.